File: qrenderdoc/Widgets/RangeHistogram.cpp
/******************************************************************************
 * The MIT License (MIT)
 *
 * Copyright (c) 2016-2019 Baldur Karlsson
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 ******************************************************************************/

#include "RangeHistogram.h"
#include <float.h>
#include <math.h>
#include <QMouseEvent>
#include <QPainter>

RangeHistogram::RangeHistogram(QWidget *parent) : QWidget(parent)
{
  setMouseTracking(true);
}

RangeHistogram::~RangeHistogram()
{
}

// sets the range of data where the histogram data was calculated.
void RangeHistogram::setHistogramRange(float min, float max)
{
  m_HistogramMin = min;
  m_HistogramMax = max;
}

// sets the minimum and maximum as well as the black and white points
void RangeHistogram::setRange(float min, float max)
{
  m_RangeMin = min;
  if(min < 0.0f)
    m_RangeMax = qMax((min - FLT_EPSILON) * (1.0f - m_MinRangeSize), max);
  else
    m_RangeMax = qMax((min + FLT_EPSILON) * (1.0f + m_MinRangeSize), max);

  m_BlackPoint = m_RangeMin;
  m_WhitePoint = m_RangeMax;

  update();
  emit rangeUpdated();
}

bool RangeHistogram::ValidRange() const
{
  if(qIsInf(m_WhitePoint) || qIsNaN(m_WhitePoint) || qIsInf(m_BlackPoint) || qIsNaN(m_BlackPoint) ||
     qIsInf(m_RangeMax) || qIsNaN(m_RangeMax) || qIsInf(m_RangeMin) || qIsNaN(m_RangeMin) ||
     qIsInf(m_RangeMax - m_RangeMin) || qIsNaN(m_RangeMax - m_RangeMin) ||
     qIsInf(m_WhitePoint - m_BlackPoint) || qIsNaN(m_WhitePoint - m_BlackPoint))
  {
    return false;
  }

  return true;
}

void RangeHistogram::setHistogramData(const QVector<uint32_t> &histogram)
{
  m_HistogramData = histogram;

  update();
}

void RangeHistogram::setBlackPoint(float val)
{
  if(val <= m_RangeMin)
    m_BlackPoint = m_RangeMin = val;
  else
    m_BlackPoint = val;

  update();
  emit rangeUpdated();
}

void RangeHistogram::setWhitePoint(float val)
{
  if(val >= m_RangeMax)
    m_WhitePoint = m_RangeMax = val;
  else
    m_WhitePoint = val;

  update();
  emit rangeUpdated();
}

float RangeHistogram::blackDelta()
{
  if(!ValidRange())
    return 0.0f;

  return delta(m_BlackPoint);
}

void RangeHistogram::setBlackDelta(float value)
{
  setBlackPoint(qMin(m_WhitePoint - m_MinRangeSize, value * (m_RangeMax - m_RangeMin) + m_RangeMin));
}

float RangeHistogram::whiteDelta()
{
  if(!ValidRange())
    return 0.0f;

  return delta(m_WhitePoint);
}

void RangeHistogram::setWhiteDelta(float value)
{
  setWhitePoint(qMax(m_BlackPoint + m_MinRangeSize, value * (m_RangeMax - m_RangeMin) + m_RangeMin));
}

float RangeHistogram::delta(float val) const
{
  return (val - m_RangeMin) / (m_RangeMax - m_RangeMin);
}

void RangeHistogram::mousePressEvent(QMouseEvent *e)
{
  if(e->button() != Qt::LeftButton || !ValidRange())
    return;

  QRect r = rect();

  int whiteX = (int)(whiteDelta() * r.width());
  int blackX = (int)(blackDelta() * r.width());

  QPointF whiteVec(whiteX - e->pos().x(), rect().height() - e->pos().y());
  QPointF blackVec(blackX - e->pos().x(), e->pos().y());

  float whitedist = (float)sqrt(whiteVec.x() * whiteVec.x() + whiteVec.y() * whiteVec.y());
  float blackdist = (float)sqrt(blackVec.x() * blackVec.x() + blackVec.y() * blackVec.y());

  if(whitedist < blackdist && whitedist < 18.0f)
    m_DragMode = DraggingMode::White;
  else if(blackdist < whitedist && blackdist < 18.0f)
    m_DragMode = DraggingMode::Black;
  else if(e->pos().x() > whiteX)
    m_DragMode = DraggingMode::White;
  else if(e->pos().x() < blackX)
    m_DragMode = DraggingMode::Black;

  if(m_DragMode == DraggingMode::White)
  {
    float newWhite = (float)(e->pos().x() - totalSpace()) / (float)regionWidth();

    setWhiteDelta(qBound(blackDelta() + m_MinRangeSize, newWhite, 1.0f));
  }
  else if(m_DragMode == DraggingMode::Black)
  {
    float newBlack = (float)(e->pos().x() - totalSpace()) / (float)regionWidth();

    setBlackDelta(qBound(0.0f, newBlack, whiteDelta() - m_MinRangeSize));
  }

  emit rangeUpdated();

  if(m_DragMode != DraggingMode::None)
    update();

  m_MousePrev = e->pos();
}

void RangeHistogram::mouseReleaseEvent(QMouseEvent *e)
{
  m_DragMode = DraggingMode::None;

  m_MousePrev = QPoint(-1, -1);
}

void RangeHistogram::mouseMoveEvent(QMouseEvent *e)
{
  QPoint pos = e->pos();
  if(ValidRange() && (e->buttons() & Qt::LeftButton) && pos != m_MousePrev)
  {
    if(m_DragMode == DraggingMode::White)
    {
      float newWhite = (float)(pos.x() - totalSpace()) / (float)regionWidth();

      setWhiteDelta(qBound(blackDelta() + m_MinRangeSize, newWhite, 1.0f));
    }
    else if(m_DragMode == DraggingMode::Black)
    {
      float newBlack = (float)(pos.x() - totalSpace()) / (float)regionWidth();

      setBlackDelta(qBound(0.0f, newBlack, whiteDelta() - m_MinRangeSize));
    }

    emit rangeUpdated();

    if(m_DragMode != DraggingMode::None)
      update();

    m_MousePrev = pos;
  }
}

void RangeHistogram::paintEvent(QPaintEvent *e)
{
  QPainter p(this);

  p.setRenderHint(QPainter::Antialiasing);

  const QBrush blackBrush(QColor(0, 0, 0));
  const QBrush redBrush(QColor(60, 0, 0));
  const QBrush greenBrush(QColor(0, 128, 0));
  const QBrush whiteBrush(QColor(255, 255, 255));

  QRectF r = rect();

  p.eraseRect(r);

  r = r.marginsRemoved(QMarginsF(m_Margin, m_Margin, m_Margin, m_Margin));

  p.fillRect(r, palette().brush(QPalette::Shadow));

  QMarginsF border(m_Border, m_Border, m_Border, m_Border);
  border /= devicePixelRatioF();

  r = r.marginsRemoved(border);

  p.fillRect(r, ValidRange() ? palette().brush(QPalette::Inactive, QPalette::Highlight) : redBrush);

  int whiteX = (int)(whiteDelta() * r.width());
  int blackX = (int)(blackDelta() * r.width() + 0.5);

  QRectF blackPoint(r.topLeft(), QSize(blackX, r.height()));
  QRectF whitePoint(r.left() + whiteX, r.top(), r.width() - whiteX, r.height());

  if(ValidRange())
  {
    p.setPen(QPen(palette().color(QPalette::Dark)));
    p.drawLine(blackPoint.topRight(), blackPoint.bottomRight());
    p.drawLine(whitePoint.topLeft(), whitePoint.bottomLeft());
  }

  p.fillRect(whitePoint, whiteBrush);
  p.fillRect(blackPoint, blackBrush);

  if(!ValidRange())
    return;

  if(!m_HistogramData.isEmpty())
  {
    float minx = delta(m_HistogramMin);
    float maxx = delta(m_HistogramMax);

    uint32_t maxval = 0;
    for(int i = 0; i < m_HistogramData.count(); i++)
    {
      float x = (float)i / (float)m_HistogramData.count();

      float xdelta = minx + x * (maxx - minx);

      if(xdelta >= 0.0f && xdelta <= 1.0f)
      {
        maxval = qMax(maxval, m_HistogramData[i]);
      }
    }

    if(maxval == 0)
      maxval = 1;

    for(int i = 0; i < m_HistogramData.count(); i++)
    {
      float x = (float)i / (float)m_HistogramData.count();
      float y = (float)m_HistogramData[i] / (float)maxval;

      float xdelta = minx + x * (maxx - minx);

      if(xdelta >= 0.0f && xdelta <= 1.0f)
      {
        float segwidth = qMax(r.width() * (maxx - minx) / (float)m_HistogramData.count(), 1.0);

        QRectF barRect(QPointF(r.left() + r.width() * (minx + x * (maxx - minx)),
                               r.bottom() - r.height() * y + 1),
                       QSizeF(segwidth, r.height() * y));

        p.fillRect(barRect, greenBrush);
      }
    }
  }

  QVector<QPointF> blackTriangle = {QPoint(blackPoint.right(), m_MarkerSize * 2),
                                    QPoint(blackPoint.right() + m_MarkerSize, 0),
                                    QPoint(blackPoint.right() - m_MarkerSize, 0)};

  QPainterPath blackPath;
  blackPath.addPolygon(QPolygonF(blackTriangle));
  p.fillPath(blackPath, palette().brush(QPalette::Dark));

  QVector<QPointF> whiteTriangle = {
      QPoint(whitePoint.left(), whitePoint.bottom() - m_MarkerSize * 2 + m_Margin),
      QPoint(whitePoint.left() + m_MarkerSize, whitePoint.bottom() + m_Margin),
      QPoint(whitePoint.left() - m_MarkerSize, whitePoint.bottom() + m_Margin)};

  QPainterPath whitePath;
  whitePath.addPolygon(QPolygonF(whiteTriangle));
  p.fillPath(whitePath, palette().brush(QPalette::Dark));

  blackTriangle[0] -= QPointF(0.0, 2.0) / devicePixelRatioF();
  blackTriangle[1] += QPointF(-2.0, 1.0) / devicePixelRatioF();
  blackTriangle[2] += QPointF(2.0, 1.0) / devicePixelRatioF();

  blackPath = QPainterPath();
  blackPath.addPolygon(QPolygonF(blackTriangle));

  whiteTriangle[0] += QPointF(0.0, 2.0) / devicePixelRatioF();
  whiteTriangle[1] -= QPointF(2.0, 1.0) / devicePixelRatioF();
  whiteTriangle[2] += QPointF(2.0, -1.0) / devicePixelRatioF();

  whitePath = QPainterPath();
  whitePath.addPolygon(QPolygonF(whiteTriangle));

  p.fillPath(blackPath, blackBrush);
  p.fillPath(whitePath, whiteBrush);
}

File: renderdocui/Controls/RangeHistogram.cs
/******************************************************************************
 * The MIT License (MIT)
 * 
 * Copyright (c) 2015-2017 Baldur Karlsson
 * Copyright (c) 2014 Crytek
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 ******************************************************************************/


using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Drawing;
using System.Data;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using renderdocui.Code;

namespace renderdocui.Controls
{
    [Designer(typeof(System.Windows.Forms.Design.ControlDesigner))]
    public partial class RangeHistogram : UserControl
    {
        public RangeHistogram()
        {
            InitializeComponent();
        }

        #region Events

        private static readonly object RangeUpdatedEvent = new object();
        public event EventHandler<RangeHistogramEventArgs> RangeUpdated
        {
            add { Events.AddHandler(RangeUpdatedEvent, value); }
            remove { Events.RemoveHandler(RangeUpdatedEvent, value); }
        }
        protected virtual void OnRangeUpdated(RangeHistogramEventArgs e)
        {
            EventHandler<RangeHistogramEventArgs> handler = (EventHandler<RangeHistogramEventArgs>)Events[RangeUpdatedEvent];
            if (handler != null)
                handler(this, e);
        }

        #endregion

        #region Privates

        private float m_RangeMax = 1.0f;
        private float m_RangeMin = 0.0f;

        private float m_WhitePoint = 1.0f;
        private float m_BlackPoint = 0.0f;

        private float m_MinRangeSize = 0.001f;

        private int m_Margin = 4;
        private int m_Border = 1;
        private int m_MarkerSize = 6;

        #endregion

        #region Code Properties

        private uint[] m_HistogramData = null;
        private float m_HistogramMin = 0.0f;
        private float m_HistogramMax = 1.0f;

        // sets the range of data where the histogram data was calculated.
        public void SetHistogramRange(float min, float max)
        {
            m_HistogramMin = min;
            m_HistogramMax = max;
        }

        // sets the minimum and maximum as well as the black and white points
        public void SetRange(float min, float max)
        {
            m_RangeMin = min;
            if (min < 0.0f)
                m_RangeMax = Math.Max((min - float.Epsilon) * (1.0f - m_MinRangeSize), max);
            else
                m_RangeMax = Math.Max((min + float.Epsilon) * (1.0f + m_MinRangeSize), max);

            m_BlackPoint = m_RangeMin;
            m_WhitePoint = m_RangeMax;

            Invalidate();
            OnRangeUpdated(new RangeHistogramEventArgs(BlackPoint, WhitePoint));
        }

        public bool ValidRange
        {
            get
            {
                if (float.IsInfinity(m_WhitePoint) || float.IsNaN(m_WhitePoint) ||
                    float.IsInfinity(m_BlackPoint) || float.IsNaN(m_BlackPoint) ||
                    float.IsInfinity(m_RangeMax) || float.IsNaN(m_RangeMax) ||
                    float.IsInfinity(m_RangeMin) || float.IsNaN(m_RangeMin) ||
                    float.IsInfinity(m_RangeMax - m_RangeMin) || float.IsNaN(m_RangeMax - m_RangeMin) ||
                    float.IsInfinity(m_WhitePoint - m_BlackPoint) || float.IsNaN(m_WhitePoint - m_BlackPoint))
                {
                    return false;
                }

                return true;
            }
        }

        [Browsable(false)]
        public uint[] HistogramData
        {
            get
            {
                return m_HistogramData;
            }
            set
            {
                m_HistogramData = value;
                Invalidate();
            }
        }

        // black point and white point are the currently selected black/white points, within the
        // minimum and maximum points. (ie. not 0 or 1)
        [Browsable(false)]
        public float BlackPoint
        {
            get
            {
                return m_BlackPoint;
            }
            set
            {
                if (value <= m_RangeMin)
                    m_BlackPoint = m_RangeMin = value;
                else
                    m_BlackPoint = value;

                Invalidate();
                OnRangeUpdated(new RangeHistogramEventArgs(BlackPoint, WhitePoint));
            }
        }
        [Browsable(false)]
        public float WhitePoint
        {
            get
            {
                return m_WhitePoint;
            }
            set
            {
                if (value >= m_RangeMax)
                    m_WhitePoint = m_RangeMax = value;
                else
                    m_WhitePoint = value;

                Invalidate();
                OnRangeUpdated(new RangeHistogramEventArgs(BlackPoint, WhitePoint));
            }
        }

        // range min/max are the current minimum and maximum values that can be set
        // for the black and white points
        [Browsable(false)]
        public float RangeMin
        {
            get
            {
                return m_RangeMin;
            }
        }
        [Browsable(false)]
        public float RangeMax
        {
            get
            {
                return m_RangeMax;
            }
        }

        #endregion

        #region Designer Properties

        [Description("The smallest possible range that can be selected"), Category("Behavior")]
        [DefaultValue(typeof(float), "0.001")]
        public float MinRangeSize
        {
            get { return m_MinRangeSize; }
            set { m_MinRangeSize = Math.Max(0.000001f, value); }
        }

        [Description("The margin around the range display"), Category("Layout")]
        [DefaultValue(typeof(int), "4")]
        public int RangeMargin
        {
            get { return m_Margin; }
            set { m_Margin = value; }
        }

        [Description("The pixel border around the range bar itself"), Category("Appearance")]
        [DefaultValue(typeof(int), "1")]
        public int Border
        {
            get { return m_Border; }
            set { m_Border = value; }
        }

        [Description("The size in pixels of each marker"), Category("Appearance")]
        [DefaultValue(typeof(int), "6")]
        public int MarkerSize
        {
            get { return m_MarkerSize; }
            set { m_MarkerSize = value; }
        }

        #endregion

        #region Internal Properties

        private int m_TotalSpace { get { return m_Margin + m_Border; } }
        private int m_RegionWidth { get { return this.Width - m_TotalSpace * 2; } }

        // these are internal only, they give [0, 1] from minimum to maximum of where the black and white points are
        private float m_BlackDelta
        {
            get
            {
                if (!ValidRange) return 0.0f;
                return GetDelta(BlackPoint);
            }
            set
            {
                BlackPoint = Math.Min(WhitePoint - MinRangeSize, value * (RangeMax - RangeMin) + RangeMin);
            }
        }
        private float m_WhiteDelta
        {
            get
            {
                if (!ValidRange) return 1.0f;
                return GetDelta(WhitePoint);
            }
            set
            {
                WhitePoint = Math.Max(BlackPoint + MinRangeSize, value * (RangeMax - RangeMin) + RangeMin);
            }
        }

        private float GetDelta(float val)
        {
            return (val - RangeMin) / (RangeMax - RangeMin);
        }

        #endregion

        #region Tooltips

        private void ShowTooltips()
        {
            blackToolTip.Show(BlackPoint.ToString("F4"), this,
                this.ClientRectangle.Left + (int)(this.ClientRectangle.Width * m_BlackDelta), this.ClientRectangle.Bottom);

            whiteToolTip.Show(WhitePoint.ToString("F4"), this,
                this.ClientRectangle.Left + (int)(this.ClientRectangle.Width * m_WhiteDelta), this.ClientRectangle.Top - 15);
        }

        #endregion

        #region Mouse Handlers

        private Point m_mousePrev = new Point(-1, -1);

        private enum DraggingMode
        {
            NONE,
            WHITE,
            BLACK,
        }
        private DraggingMode m_DragMode;

        // This handler tries to figure out which handle (white or black) you were trying to
        // grab when you clicked.
        private void RangeHistogram_MouseDown(object sender, MouseEventArgs e)
        {
            if(e.Button != MouseButtons.Left || !ValidRange)
                return;

            Rectangle rect = this.ClientRectangle;

            rect.Inflate(-m_TotalSpace, -m_TotalSpace);

            int whiteX = (int)(m_WhiteDelta * rect.Width);
            int blackX = (int)(m_BlackDelta * rect.Width);

            var whiteVec = new PointF(whiteX - e.Location.X, ClientRectangle.Height - e.Location.Y);
            var blackVec = new PointF(blackX-e.Location.X, e.Location.Y);

            float whitedist = (float)Math.Sqrt(whiteVec.X * whiteVec.X + whiteVec.Y * whiteVec.Y);
            float blackdist = (float)Math.Sqrt(blackVec.X * blackVec.X + blackVec.Y * blackVec.Y);

            System.Diagnostics.Trace.WriteLine(string.Format("white {0} black {1}", whitedist, blackdist));

            if (whitedist < blackdist && whitedist < 18.0f)
                m_DragMode = DraggingMode.WHITE;
            else if (blackdist < whitedist && blackdist < 18.0f)
                m_DragMode = DraggingMode.BLACK;
            else if (e.Location.X > whiteX)
                m_DragMode = DraggingMode.WHITE;
            else if (e.Location.X < blackX)
                m_DragMode = DraggingMode.BLACK;

            if (m_DragMode == DraggingMode.WHITE)
            {
                float newWhite = (float)(e.Location.X - m_TotalSpace) / (float)m_RegionWidth;

                m_WhiteDelta = Math.Max(m_BlackDelta + m_MinRangeSize, Math.Min(1.0f, newWhite));
            }
            else if (m_DragMode == DraggingMode.BLACK)
            {
                float newBlack = (float)(e.Location.X - m_TotalSpace) / (float)m_RegionWidth;

                m_BlackDelta = Math.Min(m_WhiteDelta - m_MinRangeSize, Math.Max(0.0f, newBlack));
            }

            OnRangeUpdated(new RangeHistogramEventArgs(BlackPoint, WhitePoint));

            if (m_DragMode != DraggingMode.NONE)
            {
                this.Invalidate();
                this.Update();
            }

            m_mousePrev.X = e.X;
            m_mousePrev.Y = e.Y;
        }

        private void RangeHistogram_MouseUp(object sender, MouseEventArgs e)
        {
            whiteToolTip.Hide(this);
            blackToolTip.Hide(this);

            m_DragMode = DraggingMode.NONE;

            m_mousePrev.X = m_mousePrev.Y = -1;
        }

        private void RangeHistogram_MouseMove(object sender, MouseEventArgs e)
        {
            if (ValidRange && e.Button == MouseButtons.Left && (e.X != m_mousePrev.X || e.Y != m_mousePrev.Y))
            {
                if (m_DragMode == DraggingMode.WHITE)
                {
                    float newWhite = (float)(e.Location.X - m_TotalSpace) / (float)m_RegionWidth;

                    m_WhiteDelta = Math.Max(m_BlackDelta + m_MinRangeSize, Math.Min(1.0f, newWhite));
                }
                else if (m_DragMode == DraggingMode.BLACK)
                {
                    float newBlack = (float)(e.Location.X - m_TotalSpace) / (float)m_RegionWidth;

                    m_BlackDelta = Math.Min(m_WhiteDelta - m_MinRangeSize, Math.Max(0.0f, newBlack));
                }

                OnRangeUpdated(new RangeHistogramEventArgs(BlackPoint, WhitePoint));

                if (m_DragMode != DraggingMode.NONE)
                {
                    this.Invalidate();
                    this.Update();
                }

                m_mousePrev.X = e.X;
                m_mousePrev.Y = e.Y;

                ShowTooltips();
            }
        }

        private void RangeHistogram_MouseLeave(object sender, EventArgs e)
        {
            whiteToolTip.Hide(this);
            blackToolTip.Hide(this);
        }

        private void RangeHistogram_MouseEnter(object sender, EventArgs e)
        {
            ShowTooltips();
        }

        #endregion

        #region Other Handlers

        private void RangeHistogram_Paint(object sender, PaintEventArgs e)
        {
            Rectangle rect = this.ClientRectangle;

            e.Graphics.FillRectangle(SystemBrushes.Control, rect);

            rect.Inflate(-m_Margin, -m_Margin);

            e.Graphics.FillRectangle(SystemBrushes.ControlText, rect);

            rect.Inflate(-m_Border, -m_Border);

            e.Graphics.FillRectangle(ValidRange ? Brushes.DarkGray : Brushes.DarkRed, rect);

            int whiteX = (int)(m_WhiteDelta * rect.Width);
            int blackX = (int)(m_BlackDelta * rect.Width);

            Rectangle blackPoint = new Rectangle(rect.Left, rect.Top, blackX, rect.Height);
            Rectangle whitePoint = new Rectangle(rect.Left + whiteX, rect.Top, rect.Width - whiteX, rect.Height);

            e.Graphics.FillRectangle(Brushes.White, whitePoint);
            e.Graphics.FillRectangle(Brushes.Black, blackPoint);

            if (!ValidRange)
                return;

            if (HistogramData != null)
            {
                float minx = GetDelta(m_HistogramMin);
                float maxx = GetDelta(m_HistogramMax);

                UInt32 maxval = UInt32.MinValue;
                for (int i = 0; i < HistogramData.Length; i++)
                {
                    float x = (float)i / (float)HistogramData.Length;

                    float xdelta = minx + x * (maxx - minx);

                    if (xdelta >= 0.0f && xdelta <= 1.0f)
                    {
                        maxval = Math.Max(maxval, HistogramData[i]);
                    }
                }

                if (maxval == 0)
                    maxval = 1;

                for (int i = 0; i < HistogramData.Length; i++)
                {
                    float x = (float)i / (float)HistogramData.Length;
                    float y = (float)HistogramData[i] / (float)maxval;

                    float xdelta = minx + x * (maxx - minx);

                    if (xdelta >= 0.0f && xdelta <= 1.0f)
                    {
                        float segwidth = Math.Max(rect.Width * (maxx - minx) / (float)HistogramData.Length, 1);

                        RectangleF barRect = new RectangleF(new PointF(rect.Left + rect.Width * (minx + x * (maxx - minx)), rect.Bottom - rect.Height * y),
                                                            new SizeF(segwidth, rect.Height * y));

                        e.Graphics.FillRectangle(Brushes.Green, barRect);
                    }
                }
            }

            Point[] blackTriangle = { new Point(blackPoint.Right, m_MarkerSize*2),
                                      new Point(blackPoint.Right+m_MarkerSize, 0),
                                      new Point(blackPoint.Right-m_MarkerSize, 0) };

            e.Graphics.FillPolygon(Brushes.DarkGray, blackTriangle);

            Point[] whiteTriangle = { new Point(whitePoint.Left, whitePoint.Bottom-m_MarkerSize*2+m_Margin),
                                      new Point(whitePoint.Left+m_MarkerSize, whitePoint.Bottom+m_Margin),
                                      new Point(whitePoint.Left-m_MarkerSize, whitePoint.Bottom+m_Margin) };

            e.Graphics.FillPolygon(Brushes.DarkGray, whiteTriangle);

            blackTriangle[0].Y -= 2;
            blackTriangle[1].Y += 1;
            blackTriangle[2].Y += 1;

            blackTriangle[1].X -= 2;
            blackTriangle[2].X += 2;

            whiteTriangle[0].Y += 2;
            whiteTriangle[1].Y -= 1;
            whiteTriangle[2].Y -= 1;

            whiteTriangle[1].X -= 2;
            whiteTriangle[2].X += 2;

            e.Graphics.FillPolygon(Brushes.Black, blackTriangle);
            e.Graphics.FillPolygon(Brushes.White, whiteTriangle);
        }

        #endregion
    }

    public class RangeHistogramEventArgs : EventArgs
    {
        private float m_black, m_white;

        public RangeHistogramEventArgs(float black, float white)
        {
            m_black = black;
            m_white = white;
        }

        public float BlackPoint
        {
            get { return m_black; }
        }

        public float WhitePoint
        {
            get { return m_white; }
        }
    }

}
